### ISSUE TITLE
The `core.Logger#setLevel` method should work like `Configurator#setLevel`

### ISSUE THREAD

**Issue by ppkarwasz** [apache/logging-log4j2#2281]
The main difference between the [`core.Logger#setLevel`](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/Logger.html#setLevel(org.apache.logging.log4j.Level)) method and [`Configurator#setLevel`](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/config/Configurator.html#setLevel(org.apache.logging.log4j.Logger,org.apache.logging.log4j.Level)) is that the former changes only the level of a specific logger, while the latter also changes the level of child loggers.

As documented in its Javadoc, the `core.Logger#setLevel` method is not part of the public API (so it can do whatever it wants). However it is used in `log4j-1.2-api` and `log4j-jul` instead of the more correct `Configurator#setLevel` method and it causes those modules to misbehave.

We can therefore:

- either change the way `Logger#setLevel` works to match the behavior of `Configurator#setLevel` (IMHO it is the best option, since it prevents mistakes from committers and users),
- or simply replace `Logger#setLevel` with `Configurator#setLevel` in `log4j-1.2-api` and `log4j-jul`.

**Comment by garydgregory** [comment]
I bitten by this issue in Apache Commons Logging

**Comment by vy** [comment]
> The main difference between the [`core.Logger#setLevel`](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/Logger.html#setLevel(org.apache.logging.log4j.Level)) method and [`Configurator#setLevel`](https://logging.apache.org/log4j/2.x/javadoc/log4j-
> We can therefore:
> 
> 1. either change the way `Logger#setLevel` works to match the behavior of `Configurator#setLevel` (IMHO it is the best option, since it prevents mistakes from committers and users),
> 2. or simply replace `Logger#setLevel` with `Configurator#setLevel` in `log4j-1.2-api` and `log4j-jul`.

Thanks for investigation, @ppkarwasz. I would like to add a 3rd option to the list above:

3. do both (would be of my preference)

**Comment by garydgregory** [comment]
TY @ppkarwasz for helping out with https://github.com/apache/logging-log4j2/pull/2282, merged! 👍

**Comment by ppkarwasz** [comment]
Let's leave this issue open, until we check `log4j-jul`.

**Comment by chia7712** [comment]
> The main difference between the [core.Logger#setLevel](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/Logger.html#setLevel(org.apache.logging.log4j.Level)) method and [Configurator#setLevel](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/config/Configurator.html#setLevel(org.apache.logging.log4j.Logger,org.apache.logging.log4j.Level)) is that the former changes only the level of a specific logger, while the latter also changes the level of child loggers.

@ppkarwasz 
Sorry for making noise on this closed issue. Out of curiosity, what is the difference between `Configurator#setLevel` and `Configurator.setAllLevels` - it seems both of they can change the level of child loggers.

**Comment by chia7712** [comment]
> what is the difference between Configurator#setLevel and Configurator.setAllLevels

it seems `Configurator.setAllLevels` also calls the `setLevel` for each `LoggerConfig`. By constrast, `setLevel` can change only the programmatically logger as they don't have `LoggerConfig`

**Comment by ppkarwasz** [comment]
> > The main difference between the [core.Logger#setLevel](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/Logger.html#setLevel(org.apache.logging.log4j.Level)) method and [Configurator#setLevel](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/config/Configurator.html#setLevel(org.apache.logging.log4j.Logger,org.apache.logging.log4j.Level)) is that the former changes only the level of a specific logger, while the latter also changes the level of child loggers.
> 
> Sorry for making noise on this closed issue. Out of curiosity, what is the difference between `Configurator#setLevel` and `Configurator.setAllLevels` - it seems both of they can change the level of child loggers.

To understand the difference between the two methods, you should take a closer look at the [architecture of Log4j Core](https://logging.apache.org/log4j/2.x/manual/architecture.html). There are two types of objects:

- [`Logger`s](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/Logger.html) that you create using:
  ```java
  Logger foo = LogManager.getLogger("foo");
  Logger fooBar = LogManager.getLogger("foo.bar");
  Logger fooBarBaz = LogManager.getLogger("foo.bar.baz");
  ```
- [`LoggerConfig`s](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/config/LoggerConfig.html) that you create using the `<Logger>` elements in a configuration file, e.g.:
  ```xml
  <Logger name="foo" level="ERROR"/>
  <Logger name="foo.bar.baz" level="WARN"/>
  ```

Multiple `Logger`s can be attached to a single `LoggerConfig`. In the example above the "foo" and "foo.bar" loggers are both attached to the "foo" logger configuration, while the "foo.bar.baz" logger is attached to the "foo.bar.baz" logger configuration.
The `Configurator` API only works on logger configurations:

- if you call `Configurator.setLevel("foo", Level.INFO)`, you'll modify the effective level of the loggers attached to the "foo" `LoggerConfig` (i.e. the "foo" and "foo.bar" loggers).
- if you call `Configurator.setAllLevels("foo", Level.INFO)`, you'll modify both the "foo" and "foo.bar.baz" `LoggerConfig`s. This will change the effective level of the "foo", "foo.bar" and "foo.bar.baz" loggers.

The situation becomes slightly more complex if you take into account the **inheritance** of levels between `LoggerConfig`s. If the "foo.bar.baz" logger configuration has a `null` level, it inherits the effective level from its parent (the "foo" `LoggerConfig`). In this case, calling `Configurator.setLevel("foo", Level.INFO)` will also change the effective level of the "foo.bar.baz" logger.

> By constrast, `setLevel` can change only the programmatically logger as they don't have `LoggerConfig`.

If a logger (like "foo.bar" above) does not have a `LoggerConfig` with the same name, both `Configurator.setLevel()` and `Configurator.setAllLevels()` will create a new `LoggerConfig`.

**Comment by chia7712** [comment]
@ppkarwasz You are the best! Thanks for this great explanation!

**Comment by vy** [comment]
@ppkarwasz, your explanation is awesome! :bow: Shouldn't we share this (maybe in a `%collapsible`) in [How do I set a logger level programmatically?](https://logging.apache.org/log4j/2.x/faq.html#reconfig_level_from_code)

**Comment by ppkarwasz** [comment]
I am not a big fan of `Configurator`, since its usage is somehow limited in multi-user environments: each `Configurator` method tries to determine the correct logger context to use.

I would prefer to redirect users to [`apache/logging-admin`](https://github.com/apache/logging-admin) once it is ready.
